### PR TITLE
Fix forward declarations & replace fwd header files

### PR DESCRIPTION
**Pull request by cngzhnp**
I had done before in #13623 and these are new fixes for other files as well. I try to use forward declaration files as much as possible. Please give feedback about if new files need to be created.

### CODE REVIEW

File: toolsrc/include/vcpkg/binarycaching.h [hunk 1]
```diff
@@ -1,16 +1,12 @@
 #pragma once
 
+#include <vcpkg/fwd/dependencies.h>
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/binarycaching.h:3]: 👍

File: toolsrc/include/vcpkg/paragraphs.h [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
 #include <vcpkg/base/expected.h>
 
 #include <vcpkg/binaryparagraph.h>
-#include <vcpkg/paragraphparser.h>
-#include <vcpkg/vcpkgpaths.h>
+
+namespace vckpg::Parse
+{
+    struct ParseControlErrorInfo;
+}
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/paragraphs.h:12]: why is this not in `vcpkg/fwd/paragraphparser.h`?

**Review by strega-nil — CHANGES_REQUESTED** [review]
Just the one thing, otherwise this is so great 😄

File: toolsrc/include/vcpkg/paragraphs.h (hunk 2 above)
Comment (cngzhnp) [reply toolsrc/include/vcpkg/paragraphs.h:12]: Escaped my notice, I will add a forward declaration file for this :)

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
(Approval obviously conditioned on a successful build :) )

**Comment by BillyONeal** [comment]
@cngzhnp It got your changes with out azp run : https://dev.azure.com/vcpkg/public/_build/results?buildId=43989&view=results

**Comment by cngzhnp** [comment]
@BillyONeal thanks, maybe not a good place to mention or ask but if the azure-pipelines bot can do format with commands here, it would be easier for us. I always forget to run the format tool before committing the changes :)

**Comment by BillyONeal** [comment]
Yeah, it's something we've wanted to do; unfortunately we're stuck between a rock and a hard place here. Azure Pipelines can't write any changes back to GitHub (like the formatting changes we want), and GitHub Actions can't control Azure to create and destroy the test machines (we need bigger than the dual core boxes Actions come with).

**Comment by cngzhnp** [comment]
> Yeah, it's something we've wanted to do; unfortunately we're stuck between a rock and a hard place here. Azure Pipelines can't write any changes back to GitHub (like the formatting changes we want), and GitHub Actions can't control Azure to create and destroy the test machines (we need bigger than the dual core boxes Actions come with).

Thanks for the reply @BillyONeal. Now, it is ready to merge I guess. (condition is checked) :)

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**